commands/operator-sdk/cmd/scorecard/scorecard.go

```diff
@@ -0,0 +1,252 @@
+// Copyright 2018 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scorecard
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	k8sInternal "github.com/operator-framework/operator-sdk/internal/util/k8sutil"
+	"github.com/operator-framework/operator-sdk/internal/util/yamlutil"
+
+	olmApi "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	v1 "k8s.io/api/core/v1"
+	extscheme "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/scheme"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/discovery/cached"
+	"k8s.io/client-go/kubernetes"
+	cgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// TODO: use a config file to reduce number of flags users
+// have to provide for the config
+
+// Config stores all scorecard config passed as flags
+type Config struct {
+	Namespace      string
+	KubeconfigPath string
+	InitTimeout    int
+	CSVPath        string
+	BasicTests     bool
+	OlmTests       bool
+	TenantTests    bool
+	NamespacedMan  string
+	GlobalMan      string
+	CrMan          string
+	Verbose        bool
+}
+
+var SCConf Config
+
+const (
+	basicOperator  = "Basic Operator"
+	olmIntegration = "OLM Integration"
+	goodTenant     = "Good Tenant"
+)
+
+// TODO: add point weights to tests
+type scorecardTest struct {
+	testType      string
+	name          string
+	description   string
+	earnedPoints  int
+	maximumPoints int
+}
+
+type cleanupFn func() error
+
+var (
+	apiversion     string
+	kind           string
+	name           string
+	kubeconfig     *rest.Config
+	scTests        []scorecardTest
+	dynamicDecoder runtime.Decoder
+	runtimeClient  client.Client
+	restMapper     *restmapper.DeferredDiscoveryRESTMapper
+	deploymentName string
+	proxyPod       *v1.Pod
+	cleanupFns     []cleanupFn
+)
+
+const scorecardPodName = "operator-scorecard-test"
+
+func ScorecardTests(cmd *cobra.Command, args []string) error {
```

Comment:
Not something I'd suggest for this PR, but it might be nice to abstract some of the complexity of this function into new types and/or functions if possible, especially with it being 150+ lines.

Maybe with something like:
```go
type TestSuite struct {
  Name string
  Description string
  Enabled bool
  Tests []Test
}

type Test interface {
  Name() string
  Description() string
  MaximumPoints() int
  Run() (earnedPoints int, err error)
}
```

With this, we could create the test suites and group the tests into them up front, then execute everything in nested for/range loops. Something like this would also make it easier for us to expose these abstractions so users could hook their own tests in.

Reply:
Yeah, I was thinking about potentially doing this, as there is no easy way to get descriptions for individual tests with the way it is designed right now, which is something I want to implement later on. This would be a much cleaner design as well. I'll do this in a follow up PR after this is merged.